### PR TITLE
WIP - (FM-7596) Draft of RSAPI Transport layer

### PR DESCRIPTION
**Pull request by da-ar**
Not happy with the transport implementation being described here, I think we need an interface class.

Added file paths to the code samples, as in @DavidS notes. I think they help make things clearer.

### CODE REVIEW

File: language/resource-api/README.md [hunk 1]
```diff
@@ -333,33 +344,101 @@ When a resource is marked with `noop => true`, either locally or through a globa
...
+# lib/puppet/provider/nx9k_vlan/nexus.rb
+class Puppet::Provider::Nx9k_vlan::Nexus
   def set(context, changes, noop: false)
     changes.each do |name, change|
       is = change.has_key? :is ? change[:is] : get_single(name)
       should = change[:should]
       # ...
-      context.device.do_something unless noop
+      if context.transport.verify
+        context.transport.do_something if noop
```
Comment (Thomas-Franklin) [inline language/resource-api/README.md]: Should this not be `unless noop`?

Comment (da-ar) [reply language/resource-api/README.md]: lol - yep

File: language/resource-api/README.md [hunk 2]
```diff
@@ -333,33 +344,101 @@ When a resource is marked with `noop => true`, either locally or through a globa
...
+    hostname: {
+      type: 'String',
+      desc: 'The host to connect to.',
+    },
+    username: {
+      type: 'String',
+      desc: 'The user.',
+    },
+    password: {
+      type: 'Sensitive[String]',
```
Comment (adreyer) [inline language/resource-api/README.md]: We discussed using sensitive like this in the task spec and decided it wasn't what we wanted. This means the password field will only accept a value of type Senstitive where what we mean is this value should be treated as sensitive information. For the task spec we instead added a new key `sensitive` next to type and description to capture this. I would encourage doing something similar here.

File: language/resource-api/README.md [hunk 3]
```diff
@@ -333,33 +344,101 @@ When a resource is marked with `noop => true`, either locally or through a globa
...
+  def close; end
+end
+```
+
+The transport layer consists of the schema and the implementation. The schema is defined in the same manner as a `Type`, except instead of `attributes`,
+ there is now `connection_info` and it describes the shape of the data which should be passed to the implementation for a connection to be made.
+
+The implementation must follow the structure as defined above and provide the following methods:
+
+  * initialize(connection_info)
```
Comment (adreyer) [inline language/resource-api/README.md]: Is initialize expected to open the connection?

File: language/resource-api/README.md [hunk 4]
```diff
@@ -333,33 +344,101 @@ When a resource is marked with `noop => true`, either locally or through a globa
...
+    * `transport_info` contains validated hash matching schema
+  * facts
+    * Access the device and return a facts hash
+  * verify
+    * Perform an operation that returns true if the connection is correctly established
```
Comment (adreyer) [inline language/resource-api/README.md]: If the connection could be established but is currenlty closed should verify return false? How would the transport recover from something like a timeout? Is that up to the code using the transport?

File: language/resource-api/README.md [hunk 5]
```diff
@@ -333,33 +344,101 @@ When a resource is marked with `noop => true`, either locally or through a globa
...
+  * close
+    * Close the connection
+
+### Legacy Support
+
+Prior support for `remote_resource` was provided through `Puppet::Util::NetworkDevice`. This code can be updated by inheriting from `Puppet::ResourceApi::Transport::Forwarder` rather than `Puppet::Util::NetworkDevice::Simple::Device`.
```
Comment (adreyer) [inline language/resource-api/README.md]: Is there a reason we don't just make `Puppet::Util::NetworkDevice::Simple::Device` the forwarder?

File: language/resource-api/README.md [hunk 6]
```diff
@@ -333,33 +344,101 @@ When a resource is marked with `noop => true`, either locally or through a globa
...
+#  lib/puppet/transport/nexus.rb
+class Puppet::Transport::Nexus
+  def initialize(transport_info); end
+  def facts; end
+  def verify; end
+  def close; end
+end
+```
+
+The transport layer consists of the schema and the implementation. The schema is defined in the same manner as a `Type`, except instead of `attributes`,
```
Comment (DavidS) [inline language/resource-api/README.md]: Put this with an additional sentence what a `transport` is at the top of the chapter

File: language/resource-api/README.md (hunk 2 above)
Comment (adreyer) [reply language/resource-api/README.md]: the run_task function turns them into instances of Puppet::Sensitive and then the transports unwrap the sensitive values at the last minute.

https://github.com/puppetlabs/bolt/blob/fb186f9165103b399a6960a7fc42983c4ce714dd/bolt-modules/boltlib/lib/puppet/functions/run_task.rb#L135
https://github.com/puppetlabs/bolt/blob/fb186f9165103b399a6960a7fc42983c4ce714dd/lib/bolt/transport/ssh.rb#L110

File: language/resource-api/README.md (hunk 5 above)
Comment (adreyer) [reply language/resource-api/README.md]: ah cool this makes more sense.

File: language/resource-api/README.md [hunk 7]
```diff
@@ -333,33 +344,101 @@ When a resource is marked with `noop => true`, either locally or through a globa
...
+  * close
+    * Close the connection
+
+### Legacy Support
+
+Prior support for `remote_resource` was provided through `Puppet::Util::NetworkDevice`. This code can be updated by inheriting from `Puppet::ResourceApi::Transport::Forwarder` rather than `Puppet::Util::NetworkDevice::Simple::Device`.
+
+```ruby
+# lib/puppet/util/network_device/nexus/device.rb
+module Puppet::Util::NetworkDevice::Nexus
```
Comment (DavidS) [inline language/resource-api/README.md]: This will need a  `require 'puppet/resource_api'`

File: language/resource-api/README.md (hunk 2 above)
Comment (DavidS) [reply language/resource-api/README.md]: Added language explaining this.

File: language/resource-api/README.md (hunk 3 above)
Comment (DavidS) [reply language/resource-api/README.md]: Added more extensive explanation to the doc

File: language/resource-api/README.md (hunk 4 above)
Comment (DavidS) [reply language/resource-api/README.md]: Added more extensive explanation to the doc, and fixed up the error handling to be ruby-style, not go-style.

File: language/resource-api/README.md (hunk 6 above)
Comment (DavidS) [reply language/resource-api/README.md]: Fixed the paragraph.

**Comment by da-ar** [comment]
+1

**Comment by DavidS** [comment]
@clairecadman can you have a quick look over the changes here, and apply your magic, please?

**Comment by clairecadman** [comment]
@DavidS Yes I'll take a look this afternoon!

**Comment by clairecadman** [comment]
@DavidS I made a couple of minor edits. But line 453 needs to be fixed, I'm just not sure what it means. Thanks!

File: language/resource-api/README.md [hunk 8]
```diff
@@ -333,33 +344,166 @@ When a resource is marked with `noop => true`, either locally or through a globa
...
+2. `connect` to the transport by name, passing the connection info
+3. When the transport has been initialized, `connect` will return the `Transport` object
+
+See above for possible exceptions coming from initializing a transport.
+
+To get a list of all registered transports, call the `list` method:
+
+`Puppet::ResourceApi::Transport.list`
+
+It will return a hash of all registered transport schemas keyed by their name. Each entry in the list is the transport schema definition as passed to `register_transport`.
```
Comment (Magisus) [inline language/resource-api/README.md]: Just to be clear, does this mean each value is also a hash (exactly like the one passed to `register`)? Thinking about the API access to this, and want to figure out what kinds of data conversion we would need (or wouldn't need) to do.

File: language/resource-api/README.md [hunk 9]
```diff
@@ -333,33 +344,166 @@ When a resource is marked with `noop => true`, either locally or through a globa
...
+
+```ruby
+# lib/puppet/transport/nexus_schema.rb
+Puppet::ResourceAPI.register_transport(
+  name: 'nexus', # points at class Puppet::Transport::Nexus
+  desc: 'Connects to a Cisco Nexus device'
+  features: [], # TODO
+  connection_info: {
+    connector_type: {
+      type: 'Enum[nxapi, ssh, grpc, pidgeon]',
```
Comment (ALawley) [inline language/resource-api/README.md]: Will this be supporting the `Optional` data type?

File: language/resource-api/README.md [hunk 10]
```diff
@@ -333,33 +344,166 @@ When a resource is marked with `noop => true`, either locally or through a globa
...
+
+1. On `initialize`, the transport may apply deeper validation to the passed connection information, like mutual exclusivity of optional values (e.g. `password` OR `key`) and throw an `ArgumentError`
+1. On `initialize`, the transport may (but doesn't have to) try to establish a connection to the remote target. If this fails due to unrecoverable errors (e.g. "password rejected"), an appropriate exception will be thrown.
+
+1. `verify` and `facts`, like `initialize`, only needs to throw exceptions when unrecoverable errors are encountered, or when the retry logic times out.
+
+1. On any other operation will primarily be used by providers and tasks talking to this kind of target. Those operations are free to use any error signalling mechanism that is convenient. Providers and tasks will have to provide appropriate error signalling to the runtime.
```
[the remark this answers is not in the document]
Comment (DavidS) [reply language/resource-api/README.md]: Me neither. And I wrote it. :embarrassed: Let me try again:

> Any other methods on the transport are to be used by providers and tasks talking to this kind of target. ...

File: language/resource-api/README.md (hunk 8 above)
Comment (DavidS) [reply language/resource-api/README.md]: Yes. `list` will return a ruby hash of hashes. I'll need to improve the wording on this one a little bit to make that clearer.

The current implementation is quite primitive: https://github.com/puppetlabs/puppet-resource_api/pull/150/commits/bd3a21e0abd825f55ef928189d5ebb8a256b7c6c

File: language/resource-api/README.md (hunk 9 above)
Comment (DavidS) [reply language/resource-api/README.md]: The Resource API does not put any restrictions on the types folks can use here.

That said, I would understand if you want to add guidance to developers around what can be mapped into UI (String (including Pattern and Enum), Number, Variant, Optional, Array, sensitive) and do a generic JSON(?) input field for everything else.

**Comment by da-ar** [comment]
@DavidS reminder that you have some local changes to push.
